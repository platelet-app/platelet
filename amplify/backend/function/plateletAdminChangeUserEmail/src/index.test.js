const indexModule = require("./index");
const aws = require("aws-sdk");

jest.mock(
    "/opt/appSyncRequest",
    () => {
        return {
            request: jest.fn(),
            errorCheck: jest.fn(),
        };
    },
    { virtual: true }
);
jest.mock(
    "/opt/graphql/mutations",
    () => {
        return {
            updateUser: "updateUserMutation",
        };
    },
    { virtual: true }
);
jest.mock(
    "/opt/graphql/queries",
    () => {
        return {
            getUser: "getUserQuery",
        };
    },
    { virtual: true }
);

// mock CognitoIdentityServiceProvider class
//
const cognitoIdentityServiceProvider = Object.getPrototypeOf(
    new aws.CognitoIdentityServiceProvider()
);

const appsyncModule = require("/opt/appSyncRequest");

const userId = "userId";

describe("plateletAdminChangeUserEmail", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    test("change the user's email", async () => {
        const user = {
            id: userId,
            _version: 1,
            contact: {
                emailAddress: "originalEmail@example.com",
            },
            username: "username",
        };
        const emailAddress = "some@email.com";
        const appsyncResponse = {
            id: userId,
            _version: 2,
            username: "username",
            contact: {
                emailAddress,
            },
        };
        const cognitoSpy = jest
            .spyOn(cognitoIdentityServiceProvider, "adminUpdateUserAttributes")
            .mockReturnValue({
                promise: () => Promise.resolve(),
            });
        const cognitoListUsersSpy = jest
            .spyOn(cognitoIdentityServiceProvider, "listUsers")
            .mockReturnValue({
                promise: () => Promise.resolve({ Users: [] }),
            });
        appsyncModule.request
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        getUser: user,
                    },
                }),
            })
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        updateUser: {
                            id: userId,
                            username: "username",
                            contact: {
                                emailAddress,
                            },
                            _version: 2,
                        },
                    },
                }),
            });
        const result = await indexModule.handler({
            arguments: { userId, emailAddress },
        });
        expect(result).toEqual(appsyncResponse);
        expect(appsyncModule.request).toMatchSnapshot();
        expect(cognitoSpy).toMatchSnapshot();
        expect(cognitoListUsersSpy).toMatchSnapshot();
    });
    test("change the user's email appsync failure", async () => {
        const user = {
            id: userId,
            _version: 1,
            contact: {
                emailAddress: "originalEmail@example.com",
            },
            username: "username",
        };
        const emailAddress = "some@email.com";
        const cognitoSpy = jest
            .spyOn(cognitoIdentityServiceProvider, "adminUpdateUserAttributes")
            .mockReturnValue({
                promise: () => Promise.resolve(),
            });
        jest.spyOn(cognitoIdentityServiceProvider, "listUsers").mockReturnValue(
            {
                promise: () => Promise.resolve({ Users: [] }),
            }
        );
        appsyncModule.request
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        getUser: user,
                    },
                }),
            })
            .mockRejectedValueOnce(new Error("appsync failure"));
        await expect(
            indexModule.handler({
                arguments: { userId, emailAddress },
            })
        ).rejects.toThrow("appsync failure");
        expect(appsyncModule.request).toMatchSnapshot();
        expect(cognitoSpy).toMatchSnapshot();
    });
    test("fail changing the user's email to an existing email", async () => {
        const emailAddress = "some@email.com";
        const cognitoListUsersSpy = jest
            .spyOn(cognitoIdentityServiceProvider, "listUsers")
            .mockReturnValue({
                promise: () =>
                    Promise.resolve({ Users: [{ Username: "username" }] }),
            });
        await expect(
            indexModule.handler({
                arguments: { userId, emailAddress },
            })
        ).rejects.toThrow("Email address already in use");
        expect(cognitoListUsersSpy).toMatchSnapshot();
    });
});
