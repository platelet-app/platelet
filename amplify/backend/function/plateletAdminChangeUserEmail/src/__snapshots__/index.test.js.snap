// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plateletAdminChangeUserEmail change the user's email 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "query": "getUserQuery",
        "variables": Object {
          "id": "userId",
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 1,
            "contact": Object {
              "emailAddress": "some@email.com",
            },
            "id": "userId",
          },
        },
      },
      "https://api.example.com/graphql",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`plateletAdminChangeUserEmail change the user's email 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "UserAttributes": Array [
          Object {
            "Name": "email",
            "Value": "some@email.com",
          },
          Object {
            "Name": "email_verified",
            "Value": "true",
          },
        ],
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
  ],
}
`;

exports[`plateletAdminChangeUserEmail change the user's email 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Filter": "email = \\"some@email.com\\"",
        "Limit": 1,
        "UserPoolId": "somePoolId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
  ],
}
`;

exports[`plateletAdminChangeUserEmail change the user's email appsync failure 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "query": "getUserQuery",
        "variables": Object {
          "id": "userId",
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 1,
            "contact": Object {
              "emailAddress": "some@email.com",
            },
            "id": "userId",
          },
        },
      },
      "https://api.example.com/graphql",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`plateletAdminChangeUserEmail change the user's email appsync failure 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "UserAttributes": Array [
          Object {
            "Name": "email",
            "Value": "some@email.com",
          },
          Object {
            "Name": "email_verified",
            "Value": "true",
          },
        ],
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "UserAttributes": Array [
          Object {
            "Name": "email",
            "Value": "originalEmail@example.com",
          },
          Object {
            "Name": "email_verified",
            "Value": "true",
          },
        ],
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
  ],
}
`;

exports[`plateletAdminChangeUserEmail fail changing the user's email to an existing email 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Filter": "email = \\"some@email.com\\"",
        "Limit": 1,
        "UserPoolId": "somePoolId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
  ],
}
`;
